wrapper/exports.py

```diff
@@ -0,0 +1,84 @@
+""" Block device exports
+
+Methods for exporting block devices from a standalone wrapper instance. The UCI
+is set up to immediately run virt-v2v-wrapper with JSON input, so the easiest
+way to run a sub-process for nbdkit is to have that input trigger this module.
+So if the JSON contains an 'nbd_export_only' object, the wrapper will just run
+the export_nbd function.
+
+The initial use case is migration from OpenStack, which has a UCI container
+running inside a source migration VM. The migration destination wrapper will
+connect to the migration source host and tell it to run the UCI container with
+the nbd_export_only JSON, which will run a function in this module to manage
+the actual nbdkit exports.
+"""
+
+import logging
+import os
+import signal
+import subprocess
+import sys
+import time
+
+from .source_hosts import NBD_READY_SENTINEL, DEFAULT_TIMEOUT
+
+
+def export_nbd(port_map):
+    """
+    Start one nbdkit process for every disk in the JSON input. Intended to be
+    run inside a UCI container inside a source conversion host.
+    """
+    log_format = '%(asctime)s:%(levelname)s:' \
+        + ' %(message)s (%(module)s:%(lineno)d)'
+    logging.basicConfig(
+        level=logging.DEBUG,
+        filename='/data/virt-v2v-wrapper.log',
+        filemode='w',
+        format=log_format)
+    logger = logging.getLogger()
+    stdout_handler = logging.StreamHandler(sys.stdout)
+    stderr_handler = logging.StreamHandler(sys.stderr)
+    logger.addHandler(stdout_handler)
+    logger.addHandler(stderr_handler)
+    logging.info('Starting up, map is %s', str(port_map))
+
+    # Start one nbdkit process per disk, using the port specified in the map
+    processes = {}
+    for disk, port in port_map.items():
+        logging.info('Exporting %s over NBD, port %s', disk, str(port))
+        cmd = ['nbdkit', '--exit-with-parent', '-p', str(port), 'file', disk]
```

Comment:
Yes, there is an SSH tunnel to the source host VM, and then podman picks a port to forward to nbdkit inside the container.

Reply:
I made it so this listens on 127.0.0.1, but from the podman call instead of the nbdkit command line. I think I would prefer it on the nbdkit command line, but I haven't found a way to make it work like that from inside the container.

Reply:
That's fine since the container will probably run in it's own network (namespace).